resolver.go

```diff
@@ -128,3 +137,176 @@ func (a *serviceAddr) Network() string {
 func (a *serviceAddr) String() string {
 	return net.JoinHostPort(a.addr, strconv.Itoa(a.port))
 }
+
+// LookupServiceFunc is the signature of functions that can be used to lookup
+// service names.
+type LookupServiceFunc func(context.Context, string) ([]net.Addr, error)
+
+// The ResolverCache type provides the implementation of a caching layer for
+// service name resolutions.
+//
+// Instances of ResolverCache are save to use concurrently from multiple
+// goroutines.
+type ResolverCache struct {
+	// The maximum age of cache entries. If zero, a default value of 1 second is
+	// used.
+	Timeout time.Duration
+
+	once sync.Once
+	cmap *resolverCacheMap
+}
+
+// LookupService resolves a service name by fetching the address list from the
+// cache, or calling lookup if the name did not exist.
+func (cache *ResolverCache) LookupService(ctx context.Context, name string, lookup LookupServiceFunc) ([]net.Addr, error) {
+	cache.once.Do(cache.init)
+
+	now := time.Now()
+	timeout := cache.timeout()
+	entry, locked := cache.cmap.lookup(name, now, now.Add(timeout))
+
+	if locked {
+		// TODO: check the error type here and discard things like context
+		// cancellations and timeouts?
+		entry.addrs, entry.err = lookup(ctx, name)
+		entry.mutex.Unlock()
+	}
+
+	entry.mutex.RLock()
+	addrs, err, expireAt := entry.addrs, entry.err, entry.expireAt
+	entry.mutex.RUnlock()
+
+	// To reduce the chances of getting cache misses on expired entries we
+	// prefetch the updated list of addresses when we're getting close to the
+	// expiration time. This is not a perfect solution and works when fetching
+	// the address list completes before the cleanup goroutine gets rid of the
+	// cache entry, but it has the advantage of being a fully non-blocking
+	// approach.
+	if now.After(expireAt.Add(-timeout / 10)) {
+		if atomic.CompareAndSwapUint32(&entry.lock, 0, 1) {
+			addrs, err := lookup(ctx, name)
+			exp := time.Now().Add(timeout)
+
+			entry.mutex.Lock()
+			entry.addrs = addrs
+			entry.err = err
+			entry.expireAt = exp
+			entry.mutex.Unlock()
+
+			atomic.StoreUint32(&entry.lock, 0)
+		}
+	}
+
+	return addrs, err
+}
+
+func (cache *ResolverCache) init() {
+	cache.cmap = &resolverCacheMap{
+		entries: make(map[string]*resolverCacheEntry),
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	runtime.SetFinalizer(cache, func(_ *ResolverCache) { cancel() })
+
+	interval := cache.timeout() / 2
+	go cache.cmap.autoDeleteExpired(ctx, interval)
+}
+
+func (cache *ResolverCache) timeout() time.Duration {
```

Comment:
We could, be I have a couple of concerns about it:
- if we update `Resolver.Timeout` then it could be confusing to the application why the value was modified
- if we cache the value and the application attempts to change it later it will have no effect (not saying it's good design, but it's not illegal if you manage concurrency properly)

The cost is pretty small so I'd rather make the code more flexible, especially because compare to an HTTP request to consul this is insignificant.